engine/src/parameters.rs

```diff
@@ -133,9 +133,44 @@ impl SubmitResult {
 #[derive(BorshSerialize, BorshDeserialize)]
 pub struct FunctionCallArgs {
     pub contract: RawAddress,
+    /// Wei compatible Borsh-encoded value field to attach an ETH balance to the transaction
+    pub value: WeiU256,
     pub input: Vec<u8>,
 }
 
+/// Legacy Borsh-encoded parameters for the `call` function, to provide backward type compatibility
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct FunctionCallArgsLegacy {
+    pub contract: RawAddress,
+    pub input: Vec<u8>,
+}
+
+/// Choosing type (type casting) between current or legacy Borsh-encoded parameters for passing to the `call` function, and to provide backward type compatibility
+pub enum CallArgsType
+where
+    Self: Sized,
+{
+    New(FunctionCallArgs),
+    Legacy(FunctionCallArgsLegacy),
+}
+
+/// Wrapper for handling in a generic way current and legacy Borsh-encoded parameters returned by the sdk::read_input_borsh() function call, to provide backward type compatibility
+pub fn call_args_type_wrap<T: Into<CallArgsType>>(args: T) -> impl Into<CallArgsType> { args }
```

Comment:
I would prefer we explicitly have some logic which tries to deserialize the new type and falls back on the old type in case of failure, for the sake of clarity.

It's not obvious to me that this will even work as intended; how does it decide at runtime which type it will use to deserialize?

Reply:
Yeah, sure, will change it!
Will make PR for toolchain update from feature branch made previously here: [toolchain-and-workspace-update](https://github.com/aurora-is-near/aurora-engine/tree/toolchain-and-workspace-update)

Reply:
Yeah, I'm also not sure about how it'll be parsed, i.e. the linear bytes taken from memory through SDK function call - this trick with impl Traits works pretty prefect with high-level types, but not with low level types, like byte sequences, I guess. So, yeah, should do this explicitly in a low level style.

Reply:
Please notice:
```
pub fn call_args_type_wrap<T: Into<CallArgsType>>(args: T) -> impl Into<CallArgsType> { args } 
```
I understand that you like `impl Trait` notation. But if you use trait bounds, you simply can use it:
```
pub fn call_args_type_wrap<T: Into<CallArgsType>>(args: T) -> T { args } 
```

Reply:
Also, it's not obvious what that function does. I think more explicit logic will be helpful.

Reply:
It was just a workaround for:
`let x: impl Foo = bar();`
'Cause Rust still doesn't support `impl Trait` in variable type declarations.
Thus, the workaround would look like:
`fn wrapper<T: Foo>(x: T) -> impl Foo { x }
let x = wrapper(bar());`
I've start use some impl Trait acrobatics, which suits more for type casting, but in this case I should use just explicit deserialization from byte sequence and try from slice (Michael gave a hint above already)...
So, I already rewrote it. Will make new PR with cherry picking of changes to new feature branch based on updated/rebased yesterday develop branch.